runtime/runtime/src/lib.rs

```diff
@@ -283,138 +339,172 @@ impl Runtime {
         debug!(target: "runtime", "{}", log_str);
     }
 
-    fn parallel_validate_transactions(
-        config: &RuntimeConfig,
+    /// Processes a group of transactions (same signer_id and public_key), sorted by ascending nonce.
+    /// Validates, calculates cost, and performs checks sequentially.
+    /// If a transaction fails validation or verification, it invokes handle_invalid_transaction
+    /// and, if `RelaxedChunkValidation` is enabled, skips to the next transaction without failing
+    /// the whole group. The resulting state (`final_state`) contains `Account` and `AccessKey` after
+    /// the last successful transaction in the group to be committed to the `state_update`.
+    /// Returns a `BatchResult` containing verified transactions, final state, and signer_id.
+    fn process_grouped_transactions(
+        &self,
+        group: TransactionBatch,
+        apply_state: &ApplyState,
+        state_update: &TrieUpdate,
         gas_price: Balance,
-        signed_txs: impl IntoParallelIterator<Item = SignedTransaction>,
+        block_height: BlockHeight,
         current_protocol_version: ProtocolVersion,
-    ) -> Vec<(CryptoHash, Result<(ValidatedTransaction, TransactionCost), InvalidTxError>)> {
-        signed_txs
-            .into_par_iter()
-            .map(|signed_tx| {
-                (
-                    signed_tx.get_hash(),
-                    match validate_transaction(config, signed_tx, current_protocol_version) {
-                        Ok(validated_tx) => {
-                            match tx_cost(
-                                config,
-                                &validated_tx,
-                                gas_price,
-                                current_protocol_version,
-                            ) {
-                                Ok(cost) => Ok((validated_tx, cost)),
-                                Err(e) => Err(InvalidTxError::from(e)),
-                            }
-                        }
-                        Err((e, _tx)) => Err(e),
-                    },
-                )
-            })
-            .collect()
+        mut temp_state: Option<(Account, AccessKey)>,
+    ) -> Result<BatchResult, RuntimeError> {
+        let first_idx = group.indices[0];
+        let signer_id = group.signed_txs[first_idx].transaction.signer_id().clone();
```

Comment:
Consider adding a method to `TransactionBatch` that abstracts away the lookup(s).

Maybe something like this even (code is written on-the-fly, might not compile off-hand):

```rust
impl TransactionBatch {
    fn iter(&self) -> impl Iterator<Item = (usize, &SignedTransaction)> {
        self.indices.iter().map(|idx| {
            (idx, self.signed_txs[idx])
        })
    }
}
```

I'm also not entirely sure why bother with these `signer_id` and `public_key` look-ups separately when you can grab the `signer_id`/`public_key` from a transaction on each iteration.